src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+        cLMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xL" + std::to_string(i), xLMap});
+    maps.insert({"cL" + std::to_string(i), cLMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xL", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cL", std::vector<int>(bits - 1, 0)});
+
+  // create c-r and x-r ladder
+  for (size_t i = max; i > 0; i--) {
+    std::map<pair<Qubit, Qubit>, int> xRMap;
+    std::map<pair<Qubit, Qubit>, int> cRMap;
+    // create the steps of the ladder
+    for (size_t j = 0; j < bits; j++) {
+      if (i > j) {
+        xRMap.insert({{j, i}, -1});
+        cRMap.insert({{i, j}, -1});
+      }
+    }
+    // save the steps in the maps
+    maps.insert({"xR" + std::to_string(i), xRMap});
+    maps.insert({"cR" + std::to_string(i), cRMap});
+  }
+  // save the complete ladder for max index evaluation
+  indices.insert({"xR", std::vector<int>(bits - 1, 0)});
+  indices.insert({"cR", std::vector<int>(bits - 1, 0)});
+
+  return std::make_pair(maps, indices);
+}
+
+// Helper function to check if the vector of a ladder step is full, meaning each
+// gate appeared in the circuit
+bool DDMinimizer::isFull(const std::vector<int>& vec) {
+  return std::all_of(vec.begin(), vec.end(),
+                     [](int value) { return value != -1; });
+}
+
+// Helper function to get the number of complete stairs in a ladder
+int DDMinimizer::getStairCount(const std::vector<int>& vec) {
+  int count = 0;
+  for (const int value : vec) {
+    if (value != -1) {
+      count++;
+    } else {
+      return count;
+    }
+  }
+  return count;
+}
+
+// Helper function to get the position of a ladder in comparison to other steps
+int DDMinimizer::getLadderPosition(const std::vector<int>& vec, int ladder) {
+  int count = 0;
+  for (size_t i = 0; i < vec.size(); i++) {
+    const int value = vec[i];
+
+    if (i == vec.size() - 1 && count != 0) {
+      if (value != -1 && value < ladder) {
+        count++;
+      }
+    }
+  }
+  return count;
+}
+
+// Helper functions to manipulate the Layout: reverse, rotate right and left
+
+std::vector<Qubit> DDMinimizer::reverseLayout(std::vector<Qubit> layout) {
+  std::vector<Qubit> reversedLayout(layout.size());
+  for (std::size_t i = 0; i < layout.size(); i++) {
+    reversedLayout[i] = layout[layout.size() - 1 - i];
+  }
+  return reversedLayout;
+}
+
+std::vector<Qubit> DDMinimizer::rotateRight(std::vector<Qubit> layout,
+                                            int stairs) {
+  const std::vector<Qubit>::size_type size = layout.size();
+  std::vector<Qubit> rotatedLayout(size);
+  for (int r = 0; r < stairs; ++r) {
+    if (!layout.empty()) {
+      rotatedLayout[size - static_cast<std::vector<Qubit>::size_type>(r + 1)] =
+          layout[static_cast<std::vector<Qubit>::size_type>(r)];
+    }
+  }
+  const std::vector<Qubit>::size_type left =
+      size - static_cast<std::vector<Qubit>::size_type>(stairs);
```

Comment:
```suggestion
  const std::size_t left = size - stairs;
```
To make this change work, you need to change the type of the variable `stairs` to `std::size_t`.